src/rdkafka_transport.c

```diff
@@ -663,6 +669,128 @@ static int rd_kafka_transport_ssl_passwd_cb (char *buf, int size, int rwflag,
 	return pwlen;
 }
 
+int verify_broker_callback(X509_STORE_CTX* ctx, void* arg)
+{
+    int validate = 0;
+    char errstr[512];
+    rd_kafka_t* rk = NULL;
+
+    errstr[0] = '\0';
+    if (arg) {
+        rk = arg;
+
+        X509* cert = NULL;
+        rd_kafka_dbg(rk, SECURITY, "SSL",
+            "Verification of peer certificate.");
+#ifndef __mips__
+        if(ctx)
+            cert = ctx->cert;
+#endif
+        if (cert) {
+            int len;
+            unsigned char* buf = NULL;
+
+            char *subj = X509_NAME_oneline(X509_get_subject_name(cert), NULL, 0);
+            char *issuer = X509_NAME_oneline(X509_get_issuer_name(cert), NULL, 0);
+
+            if (subj && issuer) {
+                rd_kafka_dbg(rk, SECURITY, "SSL", 
+                    "Validating certificate Subject: %s Issurer: %s", subj, issuer);
+            }
+
+            if (subj)
+                OPENSSL_free(subj);
+            if (issuer)
+                OPENSSL_free(issuer);
+
+            len = i2d_X509(cert, &buf);
+            rd_kafka_dbg(rk, SECURITY, "SSL",
+                "Calling client callback to verify certificate");
+
+            if (!rk->rk_conf.ssl.cert_verify_cb(buf, len, rk->rk_conf.opaque)) {
```

Comment:
If we do this who is responsible for allocating/freeing the memory?  If the callback just returns us a pointer then it needs to stick around long enough for us to return/log the error message.  In this case how would they know that is is safe to free that memory.  Possibility another callback but this is kind of clunky for just the reason why the validation failed.  The other option is for us to allocate some memory before we call the callback but then the callback is stuck to a specific size.  Do you have any other ideas on how this would work?

Reply:
@edenhill please comment if this is sufficient.